### DISCUSSION TITLE
Parent onhover to trigger changes in children

### DISCUSSION THREAD

**Discussion by martin-dimi** [lit/lit#3675] [Q&A]
Hey, what would be the correct way of triggering an event to all children of an element?
In my case, I want to notify all children when a base component is being hovered on (like a group hover), so that they can execute custom logic:

* BaseComponent (can be hovered)
* -- child 1
* -- child 2
* ---- child 1 of child2
* ---- child 2 of child2
...

What I've got right now is very manual:
```
@customElement("prompt-video")
export class BaseComponent extends LitElement {
  @query("#promptBase") _base;

  ...
  
  firstUpdated() {
     this._base.addEventListener("mouseenter", (e: MouseEvent) => this.hoverEvent(e, true))
    this._base.addEventListener("mouseleave", (e: MouseEvent) => this.hoverEvent(e, false))
  }

  hoverEvent(isHovering: boolean) {
    console.log(isHovering)
    const detail = {isHovering: isHovering}
    this.dispatchEvent(new CustomEvent("hovering", { detail, bubbles: true, composed: true }))
  }
}
```

And then listening for the `hovering` event from the interested children. Is there a convention for this?

PS: The dispatched event is not even being recognised by the children. Can we push events down to children? I guess what I need is a very lightweight shared state.

PS 2: Here's how a child listens to the event:
```
@customElement("someChild")
export class SomeChild extends LitElement {

  render() {
    return html`
      <div id="player" @hovering=${this.onHoverChange}>
      </div>
    `;
  }

  private onHoverChange(event: CustomEvent) {
    console.log(event) // never gets called
  }
}
```

**Comment by justinfagnani** [comment]
You can't really fire events down the DOM tree, so what you've done is one of the options. I'm not sure why children aren't seeing the event - can you post the code where they add the event listener?

One problem to solve with that approach is updating the listeners if the DOM changes, and dealing with more complex structure like slots.

You could use context (https://lit.dev/docs/data/context/) for this. With context you can provide the hoverable component itself to any consumer below it in the tree. Consumers can add event listeners and remove them if the hoverable component changes.

**Reply by martin-dimi** [reply]
Hey, I've updated the post with how I consume custom events from children. But yes, I see you point: if you can't fire events down the dom tree then my approach is invalid. 

Note, in my parent all the children are passed as `slots`. 

As for using `context`, what do you mean by "provide the hoverable component" - the parent component itself? What I tried is for the context to contain a `isHovering` flag and for the children to "subscribe" to it but with no luck. If it's trivial, do you mind posting a solution?
